### ISSUE TITLE
Parameter type inference doesn't work with built-in collections

### ISSUE THREAD

**Issue by Starwort** [microsoft/pyright#6508]
**Describe the bug**
Parameter type inference does not work with the built-in collections, instead falling back to Unknown

**Code or Screenshots**

```python
from typing import TypeVar, Generic
T = TypeVar('T')
class Foo:
    pass
class Bar(Generic[T]):
    pass
def fn(
    a=int(),
    b=list[int](),
    c=str(),
    d=set[str](),
    e=set(),
    f=dict[str, float](),
    g=Foo(),
    h=Bar[int](),
    i=Bar(),
):
    pass
reveal_type(fn)  # Type of "fn" is "(a: int = int(), b: Unknown = list[int](), c: str = str(), d: Unknown = set[str](), e: Unknown = set(), f: Unknown = dict[str, float](), g: Foo = Foo(), h: Bar[int] = Bar[int](), i: Bar[Unknown] = Bar()) -> None"
```

Type inference via default *does* work when the default is a primitive (such as `int()`) or a user-defined class (such as `Foo()` or `Bar[T]()`), but not for built-in collections (like `list()`, `list[T]()`, or `dict[K, V]()`)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version? Pylance VSCode extension, Pyright CLI 1.1.336 (occurs in both)

**Comment by erictraut** [comment]
This is intended. Pyright is intentionally conservative when inferring parameter types based on default arguments. In particular, it excludes `list`, `set` and `dict`. These types have invariant type parameters (because they're mutable), so inferring their type is prone to error. Often, a function accepts the less-restrictive, covariant forms. It also excludes `tuple`, whose inference rules are also very restrictive (i.e. produces a very narrow type) by design. The decision to exclude these types was based on experience that they lead to many false positive errors in typical Python code, so it's unlikely that we'd change this behavior. Even with its conservative rules for parameter inference, incorrect inferences sometimes result.

I'll also note that it's generally very dangerous to use `list`, `set` or `dict` expressions as argument defaults in Python because these expressions are not evaluated for each call site, but are instead evaluated only once. That means the same container will be used for all calls to the function, and this often leads to bugs.

In short, if you want your parameters to have well-defined types, you should include type annotations for them, not rely on parameter type inference.

**Comment by Starwort** [comment]
> I'll also note that it's generally very dangerous to use list, set or dict expressions as argument defaults in Python because these expressions are not evaluated for each call site, but are instead evaluated only once. That means the same container will be used for all calls to the function, and this often leads to bugs.

Yeah, I'm aware of that - I'm actually only providing a default object at type checking time (and replacing the default with `...` at run-time) with the primary goal of causing parameter inference. If this is by design, then that's fair enough - although I'll note that it *doesn't* exclude subclasses of those types (which perhaps is its own bug)
